PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,392 @@
+* CodeFlare operator
+* KubeRay
+* MCAD
+* ODH Distributed Workloads Component
+
+Lower priority components
+* CodeFlare SDK
+* MCAD dashboard
+* InstaScale
+
+
+### Test Approach
+
+#### Test levels and requirements:
+
+PR automated check:
+* Test execution speed is critical to provide fast feedback (less than 30 minutes)
+* Focusing on functional testing, avoid long running tests with lower value
+* Test composition (first execute unit tests, when unit tests pass then run e2e tests)
+* Use parallelization where possible to speed up execution
+
+Nightly build check (to be implemented once nightly builds are available):
+* Relaxed test execution speed requirements (i.e. 2 hours)
+* Running complete test suite
+* Test also integration with ODH Distributed Workloads Component
+
+Release build check:
+* Test speed is important to be able to provide a build fast (less than 1 hour):
+* Focusing on functional testing, avoid long running tests with lower value (tested by nightly checks)
+
+#### Testing levels and Type of testing:
+
+| Test type/development cycle  | Feature developer | PR automated checks  | PR review by reviewer | Nightly automated checks  | Release automated checks |
+| ---------------------------- | ----------------- | -------------------- | --------------------- | ------------------------- | ------------------------ |
+| Manual testing               | Yes               | No                   | Yes                   | No                        | No                       |
+| Unit testing                 | Yes               | Yes                  | No                    | Yes                       | Yes                      |
+| End to end testing           | Yes               | Yes                  | No                    | Yes                       | Yes                      |
+| Integration testing          | Optional          | Yes                  | No                    | Yes                       | Yes                      |
+| Upgrade testing              | Optional          | Yes                  | No                    | Yes                       | Yes                      |
+
+Glossary:
+
+Manual testing	- Manual verification of the functionality.
+
+Unit testing	- Unit tests covering the base functionality.
+
+End to end testing 	- Testing including all CodeFlare components, running in a cloud.
+
+Integration testing 	- Testing of interactions between ODH components and CodeFlare components in a cloud.
+
+Upgrade testing	- Testing of component upgrades (deploy new operator version) in a cloud.
+
+### Test Environment
+Test environment depends on resource requirements of execued tests. For unit tests we will leverage GitHub actions runner. For end to end/integration/upgrade testing we will currently use KinD cluster running on GitHub actions runner. In case the resources provided by GitHub actions runner are not sufficient we can consider alternatives like [testing farm](https://docs.testing-farm.io/general/0.1/index.html).
```

Comment:
For the time being, this means that we will effectively _not_ be supporting InstaScale with the initial test implementation. I'm fine with that, just wanted to call it out.